### ISSUE TITLE
angular build artifacts hash is unstable when using different node versions or project directories

### ISSUE THREAD

**Issue by xiaoxiangmoe** [angular/angular-cli#16402]
# 🐞 Bug report

### Command (mark with an `x`)

- [ ] new
- [x] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc

### Is this a regression?

No

### Description

Different machine and different node version build with different artifacts. (different file hash and different file content)

## 🔬 Minimal Reproduction

* install docker (my docker version is mac docker desktop community 2.1.0.5(40693))
* `git clone https://github.com/xiaoxiangmoe/angular-hash-problem.git`  (this repo is an empty repo create by ng-cli new)
* `cd angular-hash-problem`
* `docker run --rm --interactive --tty --mount type=bind,source="$(pwd)",target=/workspace node:13.3.0-alpine /bin/sh`
* `cd workspace/`
* `yarn build --prod`

it generate:

```
Generating ES5 bundles for differential loading...
ES5 bundle generation complete.

chunk {0} runtime-es2015.e59a6cd8f1b6ab0c3f29.js (runtime) 1.45 kB [entry] [rendered]
chunk {0} runtime-es5.e59a6cd8f1b6ab0c3f29.js (runtime) 1.45 kB [entry] [rendered]
chunk {2} polyfills-es2015.58725a5910daef768ca8.js (polyfills) 36 kB [initial] [rendered]
chunk {3} polyfills-es5.c7f38740d119d1d125aa.js (polyfills-es5) 125 kB [initial] [rendered]
chunk {1} main-es2015.9f2a69a2a7ce2c1f668b.js (main) 215 kB [initial] [rendered]
chunk {1} main-es5.9f2a69a2a7ce2c1f668b.js (main) 258 kB [initial] [rendered]
chunk {4} styles.0e4338761429b4eb16ac.css (styles) 0 bytes [initial] [rendered]
Date: 2019-12-10T03:43:32.228Z - Hash: 6aed27717e970881a87c - Time: 59765ms
```
when we build it in same node version in mac. it generate same result.

but if we build it in different node version in docker. it generate different hash, for example node:10.16.1-alpine

```
Generating ES5 bundles for differential loading...
ES5 bundle generation complete.

chunk {0} runtime-es2015.e59a6cd8f1b6ab0c3f29.js (runtime) 1.45 kB [entry] [rendered]
chunk {0} runtime-es5.e59a6cd8f1b6ab0c3f29.js (runtime) 1.45 kB [entry] [rendered]
chunk {2} polyfills-es2015.58725a5910daef768ca8.js (polyfills) 36 kB [initial] [rendered]
chunk {3} polyfills-es5.c7f38740d119d1d125aa.js (polyfills-es5) 125 kB [initial] [rendered]
chunk {1} main-es2015.9f2a69a2a7ce2c1f668b.js (main) 215 kB [initial] [rendered]
chunk {1} main-es5.9f2a69a2a7ce2c1f668b.js (main) 258 kB [initial] [rendered]
chunk {4} styles.0e4338761429b4eb16ac.css (styles) 0 bytes [initial] [rendered]
Date: 2019-12-10T04:00:33.003Z - Hash: 6aed27717e970881a87c - Time: 29734ms
```

---

also, if we run `yarn build` without `--prod`, we will found full directory in `main-es5.js`:

```
function _(module, exports, __webpack_require__) {
    module.exports = __webpack_require__(
    /*! /Users/zhaojinxiang/codes/angular-hash-problem/src/main.ts */
    "./src/main.ts");
    /***/
  }
```

if we move project from `~/codes/angular-hash-problem` into `~/repos/angular-hash-problem`, it will generate different artifacts

---

in conclusion:

* crossing operating system: stable in some specific version of node(node 13), unstable between node 10
* crossing node version: unstable between node 10 and node 11, stable between node 11,12 and 13.
* crossing directory: stable in `--prod` and unstable in `--prod=false`

**Comment by filipesilva** [comment]
I'm not sure we can do anything about this though... that seems to be webpack behaviour.

**Comment by xiaoxiangmoe** [comment]
@filipesilva
I am trying various configurations, hoping to find out whether it is caused by webpack or angular-cli.
The temporary solution is that all builds are performed in a specific version of the node docker image.

**Comment by alan-agius4** [comment]
Hi @xiaoxiangmoe, there are also a number of factors which can contribute to a hash change.

1) Different package manger versions
Different package manager version can cause different module hosting behaviour which leads may lead to different packages being include in the bundle.

2) Not using a frozen lock file.
When not using a frozen lock file, the packages installed on one machine may be different from another machine, because the package manager will not error out if the lock file is not up to date.

3. Non prod builds
As pointed out previously, there are absolute paths in the final bundle, which unless building in the production mode will not be removed.

**Comment by xiaoxiangmoe** [comment]
* Same package manager,same node version, with frozen lock, prod build.
* ng generate application foobar
* ng build foobar --prod --sourceMap=true
* cat dist/foobar/polyfills-es2015.6cb3f76b74a06c5e7d6f.js.map | cut -b 1-150

```
{"version":3,"sources":["webpack:///src/polyfills.ts","webpack:///Users/zhaojinxiang/yfd/test-angular-ten8/node_modules/zone.js/dist/zone-evergreen.js
```

You'll get absolute path in production mode.

**Comment by alan-agius4** [comment]
That's because Webpack will be emit absolute urls if the the path of a module is at an upper level from the context provided.

We provide the project root as a context to Webpack. Maybe, we should pass the workspace root instead.
https://github.com/angular/angular-cli/blob/a785d51a60b440fc12d5322422b4a18ddd22c766/packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/common.ts#L488

What is happening is that the context would be something like `/Users/zhaojinxiang/test-angular/projects/foobar` while the node_modules path will be `/Users/zhaojinxiang/test-angular/node_modules`, since the latter doesn't start with the former, Webpack will use the absolute path.

**Comment by alan-agius4** [comment]
@xiaoxiangmoe, @clydin brought to my attention that we already have an open issue to fix the absolutely paths in the source maps https://github.com/angular/angular-cli/issues/17046

Seeing that the above issue has more context. I suggest we close this issue in favour of the previously mentioned one.

**Comment by xiaoxiangmoe** [comment]
I also find build result are different between mac and Windows last year. Maybe I should test it in node 14 and angular 10.

**Comment by alan-agius4** [comment]
> I also find build result are different between mac and Windows last year

Most likely this because of different line endings in Windows and Mac. You’ll need to configure GIt to also use LF line endings.

**Comment by alan-agius4** [comment]
Closing as per above.

Let me know if you encounter this again.
